### DISCUSSION TITLE
Is there an idiomatic way that one can add middleware to an adapter?

### DISCUSSION THREAD

**Discussion by Evertt** [state-adapt/state-adapt#70] [Q&A]
So, I realized that if one wants to run certain logic at the beginning of multiple actions, then it would be cleaner if one could extract that logic into a kind of middleware. For example, let's say you're using StateAdapt to manage the state of a game being played, and for many actions you want to check if the game is still running before you apply the action. I would imagine that a middleware is the best way to solve this problem in a DRY way. But then I was surprised that I couldn't find a built-in feature to add middleware to an adapter.

So I wrote my own function. I won't show the implementation, but here's how one would use it:

```ts
const wordleAdapter = addMiddleware(
  // The adapter to add the middleware to
  // I can't decide whether to mutate the adapter in place or return a new one...
  wordleAdapterStart,

  // The actions to run the middleware on
  ["commitCurrentGuess", "updateCurrentGuess"],

  // The actual middleware function
  (state, next) => {
    if (state.gameStatus !== "playing") {
      return {
        ...state,
        error: WordleError.GameEnded,
      }
    }

    // next() returns a valid State object
    // also next() takes care of the correct payload being sent to the original action
    return next()
  }
)
```

Now, since there is no `addMiddleware` feature built-in, I can't help but wonder if I'm thinking about this the wrong way. And if I'm using this library wrong... So that's why I'm asking, is there another way one is supposed to extract this kind of logic?

**Comment by mfp22** [comment]
An "action" is the object emitted by a `Source` and represents an event, so I think you're talking about "reactions" (I also call them "state change functions"). 

Generally I avoid middleware because it can get hard to follow. And I haven't had anyone ask for it before now, so I'm just going to focus on this example until more use cases come up. 

When it comes to doing something or not doing something, for user events that could be done by disabling a button. For other events, it reminds me of `filter` from RxJS. You might have an event stream like `submit$`, which you can then convert into a source after filtering based on game state. I actually did something similar in a Wordle clone not too long ago [here](https://github.com/mfp22/wordle-jotai/blob/7cb2646aa79eae50c8643a288036fa727c8ba8ca/src/store.ts#L35) that I converted from Jōtai. But that's not using game state, so you could change `sources` to be a function where you can access observables of the store's current state, and filter based on that:

```typescript
  sources: {
    increment: ({ isPlaying$ }) => submit$.pipe(
      withLatestFrom(isPlaying$),
      filter(([, isPlaying]) => isPlaying),
      toSource('validSumission$'), // Might also need to map to correct payload
    ),
  },
```

RxJS logic can be reused like this:

```typescript
function withFilter(filter$: Observable<boolean>) {
  return <T>(submit$: Observable<T>) => submit$.pipe(
    withLatestFrom(filter$),
    filter(([, filterValue]) => filterValue),
    map(([t]) => t), // Map to payload
  );
}

// ...

  submit$ = new Source<void>('submit$')

// ... 

    increment: ({ isPlaying$ }) => submit$.pipe(withFilter(isPlaying$)),
```

Not perfect, but I think it's easier to follow than middleware. The action will not appear in Redux Devtools and it will be like `submit$` never emitted. 

If you do the filtering in the adapter side of things, you'll still have the action show up in Redux Devtools, but it will just show the state as the same thing as before. It becomes part of the application's event log, that that action happened, but it didn't matter. If you want that, that's fine. In that case, logic can be reused by extracting the functionality to another function and explicitly calling it from the state change function itself. Is isn't quite as DRY as middleware. But it's pretty good. 

Eventually I'm going to take a look at this kind of problem in general though. You're basically looking for a state machine. It's not too hard to get there from adapters. StateAdapt could have a function called `createMachine` that takes in an adapter, and then a mapping of statuses that can transition to each other. Like this:

```typescript
const wordleMachine = createMachine(wordleAdapter, {
  playing: {
    playing: ['commitCurrentGuess', 'updateCurrentGuess'],
    idle: ['commitCurrentGuess'],
  },
});
```

What this would do is redefine and wrap all the state change functions with this logic:
1. Check current `state.status`
2. Check the restrictions passed into `createMachine` and get the restrictions for the current status if they exist
3. If they do, make sure this state change function is among the allowed state change functions
4. Maybe set an error or something

It could also define selectors like `commitCurrentGuessEnabled` which would use `state.status` and the restrictions and return `true` if `commitCurrentGuess` is allowed. This seems convenient, but it also feels kind of wrong. The UI would need to have knowledge of downstream state changes rather than just being a simple declaration of an event. But I guess disabling a button requires similar downstream knowledge... I'm gonna have to come back to this topic. It doesn't hurt to start simple though. 

Sorry, I've been waiting to type out my thoughts on state machines for a while. I need to learn more about them and XState.

**Comment by Evertt** [comment]
Hey thank you for the comprehensive response. In response to the state machine suggestion, yeah maybe that makes sense in this specific case, but I'm more interested in the more generalized case of:

* You have some state that you're managing with an adapter
* The adapter offers state change functions to change the state in a "safe" way
* However, some state change functions are only valid to call when the current state is in a particular... state (in this case, a player may only continue guessing the word if the game is still ongoing). Or sometimes a state change function is valid to call, but only with some payloads...
* In any case, I think the adapter should be the one responsible for doing this validation and returning feedback to the consumer of the adapter if a state change function was erroneously called.
* And then yes, as a small side note, I thought it might happen that multiple state change functions need the exact same validation logic so I was looking for a way to keep that part DRY.

@mfp22, you said this hadn't come up before, I assume you meant the middleware suggestion, right? Because I'd be surprised if validation / error handling really haven't come up before. But I tried to do a search in these discussions and I couldn't find it.

I see in your Wordle code you just ignore invalid inputs. Which I guess is fine in the Wordle game. But usually, you would want to return feedback somehow that the last state change function call was invalid.

I've been doubting between throwing an error, or just setting an `error` property on the state itself. In the end I settled on setting an `error` property on the state itself, because that made it easy to then add a piece of UI that appears when that property is set. (Including a triggering a side-effect that clears the error after a couple seconds.)

I'm curious what your thoughts are on all this @mfp22

**Comment by mfp22** [comment]
Related: https://github.com/state-adapt/state-adapt/discussions/76

It turns out to make StateAdapt more tree-shakeable I need to add higher order functions the way Elf and NgRx Signal Store have done. This will enable your middleware approach. And state machines. And easy local storage. And pretty much everything. 

For localStorage, it could be 

```ts
adapt(0)(withLocalStorage())
```

In the simplest case.

Still figuring things out. But I'll publish an issue describing all the changes I'm thinking of in StateAdapt 3 probably this week.
